src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,952 @@
+      | (_, b) -> b 
+    in
+    let diff t1 t2 = Z.((const_offset t1) - (const_offset t2))
+    in
+    let cmp_z x y = 
+      let cmp_z_ref x y: int =
+        match x, y with
+        | (None, _), (None, _) -> 0
+        | (None, _), (Some _, _) -> -1
+        | (Some _, _), (None, _) -> 1
+        | (Some ii, _), (Some ij, _) -> ii - ij 
+      in
+      match x, y with
+      | (_, t1i, t2i), (_, t1j, t2j) -> 
+        let diff_e1 = cmp_z_ref t1i t1j in
+        if diff_e1 <> 0 then diff_e1 else
+          let diff_e2 = cmp_z_ref t2i t2j in
+          if diff_e2 <> 0 then diff_e2 else 
+            Z.to_int (Z.((diff t1i t2i) - (diff t1j t2j)))
+    in
+    let sort_z_by_expr zts =
+      match zts with
+      | None -> ()
+      | Some zts' -> Array.stable_sort cmp_z zts'
+    in
+    let sort_annotated ats = 
+      let cmp_annotated x y : int = 
+        match x, y with
+        | (i, _), (j, _) -> i - j
+      in
+      match ats with
+      | None -> ()
+      | Some ats' -> Array.stable_sort cmp_annotated ats'
+    in
+    let process_eq_classes zts = 
+      let is_const x =
+        match x with
+        | (_, (None, _), (None, _)) -> true
+        | _ -> false
+      in
+      let size_of_eq_class zts (start : int) : int =
+        let ref_elem = zts.(start) in
+        let remaining = (Array.length zts) - start - 1 in
+        let result = ref 0 in
+        for i = 0 to remaining do
+          let current_elem = zts.(start + i) in
+          if cmp_z ref_elem current_elem = 0 then result := !result + 1
+        done;
+        !result
+      in
+      let least_index_var_in_eq_class zts start size : int * Z.t =
+        let result = ref (0, Z.of_int 0) in 
+        match zts.(start) with
+        | (i, (_, b), (_, _)) -> result := (i, b);
+          for i = start + 1 to start + size - 1 do
+            match zts.(i) with
+            | (j, (_, b), (_, _)) ->
+              if j < fst !result then result := (j, b)
+          done;
+          !result
+      in
+      let all_are_const_in_eq_class zts start size : bool = 
+        let result = ref true in
+        for i = start to start + size - 1 do
+          if not (is_const zts.(i)) then result := false;
+        done;
+        !result
+      in
+      let assign_vars_in_const_eq_class ats zts start size least_i least_b =     
+        for i = start to start + size - 1 do
+          match zts.(i) with
+          | (ai, t1, t2) -> if Z.equal (diff t1 t2) (Z.of_int 0) then ats.(i) <- (ai, t1)
+            else
+              match t1 with
+              | (_, bj) -> ats.(i) <- (ai, (Some least_i, Z.sub bj least_b))
+        done
+      in
+      let assign_vars_in_non_const_eq_class ats zts start size least_i least_b = 
+        for i = start to start + size - 1 do
+          match zts.(i) with
+          | (ai, t1, _) -> 
+            let bj = const_offset t1 in
+            ats.(i) <- (ai, (Some least_i, Z.sub bj least_b))
+        done
+      in
+      match zts with
```

Comment:
Option.map here?

Reply:
This is a reassignment of an element in an array, therefore Option.map is not applicable.

Reply:
Why not? It seems that the function below is applied in the case of `Some x` and yields some `Some result`, while `None` is mapped to `None`.

Reply:
But I replaced the for loop with a `BatArray.modifyi`

Reply:
Because `map` generates a new list where the result is saved in. 
Does `BatArray.modifyi` suit you as well?

Reply:
I think those are two separate issues. `modifyi` is good. Nevertheless you are still doing a map on zts here (or an iter if you modify the code below to work in-place).

Reply:
Ahh now i get it, sorry I thought you meant the for loop, but you meant the lines below that. Yeah sure, I will fix both 👍 thanks for the clarification